gzip.go

```diff
@@ -63,35 +69,93 @@ func addLevelPool(level int) {
 // GzipResponseWriter provides an http.ResponseWriter interface, which gzips
 // bytes before writing them to the underlying response. This doesn't close the
 // writers, so don't forget to do that.
+// It can be configured to skip response smaller than minSize.
 type GzipResponseWriter struct {
 	http.ResponseWriter
 	index int // Index for gzipWriterPools.
 	gw    *gzip.Writer
+
+	code int // Saves the WriteHeader value.
+
+	minSize      int    // Specifed the minimum response size to gzip. Only the first Write call is checked.
+	buff         []byte // Holds the first part of the write before the reaching the minSize or the end of the write.
+	bytesWritten int    // Keep trace of the numbers of bytes written.
 }
 
 // Write appends data to the gzip writer.
 func (w *GzipResponseWriter) Write(b []byte) (int, error) {
-	// Lazily create the gzip.Writer, this allows empty bodies to be actually
-	// empty, for example in the case of status code 204 (no content).
-	if w.gw == nil {
-		w.init()
-	}
+	// Update the numbers of bytes written.
+	w.bytesWritten += len(b)
 
 	if _, ok := w.Header()[contentType]; !ok {
 		// If content type is not set, infer it from the uncompressed body.
 		w.Header().Set(contentType, http.DetectContentType(b))
 	}
-	return w.gw.Write(b)
+
+	// If the global write is bigger than the minSize, compression is used.
+	// Otherwise it save the write into a buffer and send it with the regular
+	// ResponseWriter at close.
+	if w.bytesWritten > w.minSize {
+		w.Header().Set(contentEncoding, "gzip")
+
+		// if the Content-Length is already set, then calls to Write on gzip
+		// will fail to set the Content-Length header since its already set
+		// See: https://github.com/golang/go/issues/14975
+		w.Header().Del(contentLength)
+
+		// Write the header and contentEncoding to gzip.
+		w.writeHeader()
+
+		n := 0
+		// If the buffer is not nil, it needs to be empty into the gzip writer.
+		if w.buff != nil {
+			if w.gw == nil {
+				w.init()
+			}
+
+			n1, err := w.gw.Write(w.buff)
+			if err != nil {
+				return n, err
+			}
+
+			// Saved the buffed length.
+			n = n1
+		}
+
+		w.buff = nil
+
+		if w.gw == nil {
+			w.init()
+		}
+
+		n1, err := w.gw.Write(b)
+		if err != nil {
+			return n, err
+		}
+
+		// Add the buffered length to the new write length.
+		n += n1
+		return n, err
+	}
+
+	// When arrived here, it's to save the writes into the buffer until it
+	// goes over the minSize or it closes.
+	w.buff = append(w.buff, b...)
+	return 0, nil
```

Comment:
If the response is not compressed, the caller get a response with **0** and **nil** which could looks like an error but which is not.
Because the actual write happens at close we can't return to caller the numbers of bytes written before GZIP.

I'm not sure what is the best option. 
Discussion is open. :smiley: